trace-dispatcher/doc/trace-dispatcher.md

```diff
@@ -0,0 +1,798 @@
+2. __trace presentation__ -- by __detail level__ context.
+
+Severity an detail level can be configured.
+
+## Filter context
+### Severity
+
+__Severity__ is expressed in terms of the enumeration provided by [section 6.2.1 of RFC 5424](https://tools.ietf.org/html/rfc5424#section-6.2.1):
+
+```haskell
+data SeverityS
+    = Debug | Info | Notice | Warning | Error | Critical | Alert | Emergency
+```
+
+..which ranges from minimum (`Debug`) to the maximum (`Emergency`) severity, and allows ignoring messages with severity level _below_ a configured global __severity cutoff__.
+
+The following __trace combinators__ affect __annotated severity__ of a trace:
+
+```haskell
+withSeverity :: Monad m => (a -> Severity) -> Trace m a -> Trace m a
+setSeverity  :: Monad m => Severity        -> Trace m a -> Trace m a
+
+-- Unconditional annotation:
+tracer   = setSeverity Notice trAddBlock
+
+-- Conditional annotation:
+tracer'' = withSeverity (\case
+                           IgnoreBlockOlderThanK{}          -> Warning
+                           IgnoreBlockAlreadyInVolatileDB{} -> Notice)
+                        trAddBlock
+```
+
+If the combinators are applied multiple times to a single trace, only the outermost application affects it -- the rest of them is ignored.
+
+```haskell
+traceWith (setSeverity Warning trAddBlock) (IgnoreBlockOlderThanK b)
+```
+
+In addition, the __severity context__ of a particular trace can be further overriden by configuration, at the __tracer namespace__ granularity -- which allows to put them above or below the __global severity cutoff__, effectively either enabling or disabling them.
+
+`Info` is the default __severity__, in the absence of trace context or configured severity overrides.
+
+NOTE: as en extension to the filtering severity type (`SeverityF`), a `SilenceF` constructor is defined, which encodes unconditional silencing of a particular trace -- and therefore serves as a semantic expression of the `nullTracer` functionality.
+
+### Privacy
+
+__Privacy__ annotations allows limiting __trace-outs__ that particular traces can reach.  It is expressed in terms of:
+
+```haskell
+data Privacy
+    = Confidential | Public
+```
+
+__Confidential__ privacy level means that the trace will not be externalised from the system, except via __standard output__.
+
+The annotation mechanism is similar to the one of severity:
+
+```haskell
+privately :: Trace m a -> Trace m a
+withPrivacy :: Monad m => (a -> Privacy) -> Trace m a -> Trace m a
+```
+
+`Public` is the default __privacy__, in the absence of privacy annotations.
+
+Trace privacy cannot be configured.
+
+See [Confidentiality and privacy filtering implementation](#Confidentiality-and-privacy -filtering-implementation) for a more full discussion of semantics.
+
+To further prevent occasional leaks of `Confidential` traces, all output from those traces is tagged with the `CONFIDENTIAL` keyword.
```

Comment:
Ok great, but what if we do `withPrivacy (const Public) (privately someTracer)`. Have I now made all confidential messages public? If yes, this seems prone to programmer error.

Reply:
Yes you have, but that seems to me to fall more in the category of "bad will", then to be in any way prone to an error.

Reply:
The concern seems reasonable.

I wonder if it will make sense to detect that we override a private context to be non-private -- and raise an error.

We don't really need `withPrivacy` to be dynamic -- what does our implementation experience tell us @jutaro ?

What I'm getting at, is that we could restrict the API like this:

- default to `Public` (which we already do)
- export `privately` with original semantics
- rename `publicly` to `unsafePublicly`
- define a new `publicly` to fail if the context is `Private`

I think that will help reducing occasional mistakes, @jutaro ?

Reply:
> define a new publicly to fail if the context is Private

Would you proposed function only allow you to make public things public then (a noop)? In that case it only serves as an assertion function. Renaming to `unsafePublicly` might be sufficient: making things private is always safe but making things public dangerous. Perhaps it'll be more clear as the library get's used.

Reply:
We've decided to just remove `publicly` and `privately` -- as indeed they do not appear to be necessary for the node integration.